### PR TITLE
GEOMESA-937 Use Geomesa-Convert for Ingest Tools

### PR DESCRIPTION
**Pull request by jahhulbert-ccri**
* Use the geomesa-convert framework for ingesting data via the command line tools
* Replace old ingest command line args with args for converter configuration
* Add transforms to converter framework for lists and maps
* Allow SFT to be specified in TypeSafe conf for ingest (as a string or a filename)
* Allow conf files with SFT and converter in the same file
* Configurable paths to SFT and Convert Conf objects in TypeSafe config files
* Allow string escapes in convert parser

Signed-off-by: Andrew Hulbert <andrew.hulbert@ccri.com>

### CODE REVIEW

File: geomesa-convert/geomesa-convert-common/src/test/scala/org/locationtech/geomesa/convert/common/TransformersTest.scala [hunk 1]
```diff
@@ -38,6 +38,15 @@ class TransformersTest extends Specification {
           exp.eval(Array(null)) must be equalTo "hello"
         }
 
+        "allow quoted strings" >> {
+          val exp = Transformers.parseTransform("'he\\'llo'")
```
Comment (elahrvivaz) [inline geomesa-convert/geomesa-convert-common/src/test/scala/org/locationtech/geomesa/convert/common/TransformersTest.scala]: nice

File: geomesa-tools/src/main/scala/org/locationtech/geomesa/tools/SftArgParser.scala [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+      .get
+  }
+
+  private[SftArgParser] def getSpecParser (specArg: String, nameOpt: Option[String]) : SpecParser = () => {
+    nameOpt.map[Option[SimpleFeatureType]] { featureName =>
+        Try { SimpleFeatureTypes.createType(featureName, specArg) }
+        match {
+          case Success(sft) => Some(sft)
+          case Failure(ex)  =>
+            logger.debug(s"Unable to parse sft spec from string $specArg with error ${ex.getMessage}")
```
Comment (elahrvivaz) [inline geomesa-tools/src/main/scala/org/locationtech/geomesa/tools/SftArgParser.scala]: should these be logged at a higher level than debug, and/or throw an exception? can't tell if failure is somewhat expected or not. Ditto in getConfParser below.

File: geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/geotools/SimpleFeatureTypes.scala [hunk 3]
```diff
@@ -609,16 +616,16 @@ object SimpleFeatureTypes {
...
 
-    def nonGreedyStringLiteral: Parser[String] =
-      ("\""+"""([^"\p{Cntrl}\\]|\\[\\'"bfnrt]|\\u[a-fA-F0-9]{4})*?"""+"\"").r
-
-    def nonGreedySingleQuoteLiteral: Parser[String] =
-      ( "\'" +"""([^"\p{Cntrl}\\]|\\[\\'"bfnrt]|\\u[a-fA-F0-9]{4})*?""" + "\'").r
-
-    def quotedString = nonGreedyStringLiteral | nonGreedySingleQuoteLiteral ^^ { x => x.drop(1).dropRight(1) }
-    def safeString = "([^,]+)".r
+//    def nonGreedyStringLiteral: Parser[String] =
```
Comment (elahrvivaz) [inline geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/geotools/SimpleFeatureTypes.scala]: remove commented out code

**Comment by elahrvivaz** [comment]
looks good

File: geomesa-tools/src/main/scala/org/locationtech/geomesa/tools/SftArgParser.scala (hunk 2 above)
Comment (jahhulbert-ccri) [inline geomesa-tools/src/main/scala/org/locationtech/geomesa/tools/SftArgParser.scala]: yeah i guess failure is expected...it's like it tries a list of things and if it works it accepts it and if not it moves on to the next one. but i guess that means you don't get error messages if they occur...perhaps returning the Trys and then if there is no success by either method you can print both errors.

Comment (jahhulbert-ccri) [inline geomesa-tools/src/main/scala/org/locationtech/geomesa/tools/SftArgParser.scala]: https://geomesa.atlassian.net/browse/GEOMESA-993

File: geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/text/EnhancedTokenParsers.scala [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+package org.locationtech.geomesa.utils.text
+
+import org.apache.commons.lang.StringEscapeUtils
+
+import scala.util.parsing.combinator.JavaTokenParsers
+
+trait EnhancedTokenParsers extends JavaTokenParsers {
+
+  def nonGreedyStringLiteral: Parser[String] =
+    ("\""+"""([^"\p{Cntrl}\\]|\\[\\'"bfnrt]|\\u[a-fA-F0-9]{4})*?"""+"\"").r
```
Comment (anthonyccri) [inline geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/text/EnhancedTokenParsers.scala]: Some impressive regex'ing here.

File: geomesa-convert/geomesa-convert-common/src/test/scala/org/locationtech/geomesa/convert/common/TransformersTest.scala [hunk 5]
```diff
@@ -281,5 +294,43 @@ class TransformersTest extends Specification {
...
+        res.size mustEqual 3
+        res.toList must containTheSameElementsAs(List(1,2,3))
+      }
+      "with numbers" >> {
+        val trans = Transformers.parseTransform("parseList('int', $0, '%')")
+        trans.eval(Array("1%2%a")).asInstanceOf[java.util.List[Int]] must throwAn[IllegalArgumentException]
+      }
+    }
+
+//    "parse maps" >> {
```
Comment (anthonyccri) [inline geomesa-convert/geomesa-convert-common/src/test/scala/org/locationtech/geomesa/convert/common/TransformersTest.scala]: To be filled in?

**Comment by elahrvivaz** [comment]
6c9c3bff55f42ca21f57f546be8f5c937836b08f

**Pull request closed without merging**